src/nanoarrow/nanoarrow_device_cuda.c

```diff
@@ -392,6 +417,251 @@ static ArrowErrorCode ArrowDeviceCudaArrayMove(struct ArrowDevice* device_src,
+                                             dst, hstream, &unknown_buffer_size_count);
+  if (code != NANOARROW_OK) {
+    ArrowDeviceCudaDestroyStream(hstream);
+    ArrowErrorSet(error, "ArrowDeviceCudaArrayViewCopyBuffers() failed");
+    return code;
+  }
+
+  CUstream hstream_data = NANOARROW_CUDA_DEFAULT_STREAM;
+
+  if (unknown_buffer_size_count > 0) {
+    // If buffers with unknown size were encountered, we need to copy all of their
+    // sizes to the CPU, synchronize, then copy all of the buffers once the CPU
+    // knows how many bytes from each buffer need to be copied.
+    code = ArrowDeviceCudaCreateStream(device_src, device_dst, &hstream_data, error);
+    if (code != NANOARROW_OK) {
+      ArrowDeviceCudaDestroyStream(hstream);
+      NANOARROW_CUDA_RETURN_NOT_OK(err, "cudaStreamCreate", error);
+    }
+
+    // Issue all required async copies
+    code = ArrowDeviceCudaResolveBufferSizesAsync(device_src, &src->array_view,
+                                                  hstream_data);
+    if (code != NANOARROW_OK) {
+      ArrowDeviceCudaDestroyStream(hstream);
+      ArrowErrorSet(error, "ArrowDeviceCudaResolveBufferSizesAsync() failed");
+      return code;
+    }
+
+    // Wait for them all to complete
+    err = cuStreamSynchronize(hstream);
+    if (err != CUDA_SUCCESS) {
+      ArrowDeviceCudaDestroyStream(hstream);
+      ArrowDeviceCudaDestroyStream(hstream_data);
+      NANOARROW_CUDA_RETURN_NOT_OK(err, "cuStreamSynchronize()", error);
+    }
+
+    // Issue async copies for data buffers. This will skip buffers that are already
+    // being copied.
+    unknown_buffer_size_count = 0;
+    code = ArrowDeviceCudaArrayViewCopyBuffers(device_src, &src->array_view, device_dst,
+                                               dst, hstream, &unknown_buffer_size_count);
+    if (code != NANOARROW_OK) {
+      ArrowDeviceCudaDestroyStream(hstream);
+      ArrowDeviceCudaDestroyStream(hstream_data);
+      ArrowErrorSet(error,
+                    "ArrowDeviceCudaArrayViewCopyInternal() for data buffers failed");
+      return code;
+    }
+
+    // There should not have been any buffers with unknown size in this round of copying
+    NANOARROW_DCHECK(unknown_buffer_size_count == 0);
+  }
+
+  // TODO: Populate dst->sync_event and record the copying work
+  // instead of synchronizing here.
+  if (hstream != NANOARROW_CUDA_DEFAULT_STREAM) {
+    err = cuStreamSynchronize(hstream);
+    ArrowDeviceCudaDestroyStream(hstream);
+    if (err != CUDA_SUCCESS) {
+      ArrowDeviceCudaDestroyStream(hstream_data);
+      NANOARROW_CUDA_RETURN_NOT_OK(err, "cuStreamSynchronize(hstream)", error);
+    }
+  }
+
+  if (hstream_data != NANOARROW_CUDA_DEFAULT_STREAM) {
+    err = cuStreamSynchronize(hstream_data);
+    ArrowDeviceCudaDestroyStream(hstream_data);
+    if (err != CUDA_SUCCESS) {
+      NANOARROW_CUDA_RETURN_NOT_OK(err, "cuStreamSynchronize(hstream_data)", error);
+    }
+  }
```

Comment:
If we're not populating the sync event we probably want to synchronize the stream regardless of whether it's the default stream or not